src/jit/morph.cpp

```diff
@@ -9152,6 +9152,23 @@ GenTreePtr          Compiler::fgMorphSmpOp(GenTreePtr tree, MorphAddrContext* ma
 
     ASSIGN_HELPER_FOR_MOD:
 
+        // For "val % 1", return 0 if op1 doesn't have any side effects
+        // and we are not in the CSE phase, we cannot discard 'tree'
+        // because it may contain CSE expressions that we haven't yet examined.
+        //
+        if (((op1->gtFlags & GTF_SIDE_EFFECT) == 0) && !optValnumCSE_phase)
+        {
+            if (((op2->gtOper == GT_CNS_INT) && (op2->gtIntConCommon.IconValue() == 1))
+                || ((op2->gtOper == GT_CNS_LNG) && (op2->gtIntConCommon.LngValue() == 1)))
+            {
+                GenTreePtr zeroNode = new(this, GT_CNS_INT) GenTreeIntCon(typ, 0);
```

Comment:
Either case worked for me but I didn't see any tree deletion in the pre-pass so I just followed the convention. In fact, I'm not sure deleting tree actually reclaims the data structure. If deleting the tree is the right semantic here, I will add it back.